File: Search_a_2D_Matrix_v1.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    const row = matrix.length, col = matrix[0].length
    let left = 0, right = row * col - 1
    while (left <= right) {
        const mid = Math.floor((left + right) / 2)
        const ele = matrix[Math.floor(mid / col)][mid % col]
        if (ele === target) return true
        else if (ele < target) left = mid + 1
        else right = mid - 1
    }
    return false
};

(() => {
    let testCases = [{ matrix: [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], target: 3 },
    { matrix: [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], target: 13 },
    { matrix: [[1], [3]], target: 3 }]
    for (let testCase of testCases) {
        console.log(searchMatrix(testCase.matrix, testCase.target))
    }
})()

File: Search_a_2D_Matrix_v1.kt

class Solution {
    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
        if (matrix.isEmpty()) {
            return false
        }
        val firstRow = matrix[0]
        val m = matrix.size
        val n = firstRow.size
        var low = 0
        var high = m - 1
        while (low <= high) {
            val mid = (low + high) / 2
            if (matrix[mid][0] > target) {
                high = mid - 1
            } else if (matrix[mid][0] < target) {
                low = mid + 1
            } else {
                return true
            }
        }

        if (high < 0) {
            return false
        }
        // high 指向最后一个满足 matrix[high][0] <= target 的行
        val index = high
        low = 0
        high = n - 1
        while (low <= high) {
            val mid = (low + high) / 2
            if (matrix[index][mid] > target) {
                high = mid - 1
            } else if (matrix[index][mid] < target) {
                low = mid + 1
            } else {
                return true
            }
        }
        return false
    }
}


fun main() {
    val testCase =
        arrayOf<IntArray>(
            intArrayOf(1, 3, 5, 7),
            intArrayOf(10, 11, 16, 20),
            intArrayOf(23, 30, 34, 60)
        )
    val sol = Solution()
    println(sol.searchMatrix(testCase, 11))
    println(sol.searchMatrix(testCase, 3))
    println(sol.searchMatrix(testCase, 12))
}